streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java

```diff
@@ -0,0 +1,303 @@
+    private KTable<String, Long> table;
+    final Map<String, String> results = new HashMap<>();
+    private ForeachAction<String, String> foreachAction;
+
+    @Before
+    public void before() {
+        testNo++;
+        builder = new KStreamBuilder();
+        createTopics();
+        streamsConfiguration = new Properties();
+        final String applicationId = "globalOne-table-test-" + testNo;
+        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, applicationId);
+        streamsConfiguration
+                .put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        streamsConfiguration.put(StreamsConfig.ZOOKEEPER_CONNECT_CONFIG, CLUSTER.zKConnectString());
+        streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+        streamsConfiguration.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
+        globalTable = builder.globalTable(Serdes.Long(), Serdes.String(), globalOne, globalStore);
+        stream = builder.stream(Serdes.String(), Serdes.Long(), inputStream);
+        table = builder.table(Serdes.String(), Serdes.Long(), inputTable, "table");
+        foreachAction = new ForeachAction<String, String>() {
+            @Override
+            public void apply(final String key, final String value) {
+                results.put(key, value);
+            }
+        };
+    }
+
+    @After
+    public void whenShuttingDown() throws IOException {
+        if (kafkaStreams != null) {
+            kafkaStreams.close();
+        }
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+    }
+
+    @Test
+    public void shouldKStreamGlobalKTableLeftJoin() throws Exception {
+        final KStream<String, String> streamTableJoin = stream.leftJoin(globalTable, keyMapper, joiner);
+        streamTableJoin.foreach(foreachAction);
+        runAndVerifyLeftJoin(inputStream);
+    }
+
+    @Test
+    public void shouldKStreamGlobalKTableJoin() throws Exception {
+        final KStream<String, String> streamTableJoin = stream.join(globalTable, keyMapper, joiner);
+        streamTableJoin.foreach(foreachAction);
+        runAndVerifyJoin(inputStream);
+    }
+
+    @Test
+    public void shouldKTableGlobalKTableLeftJoin() throws Exception {
+        final KTable<String, String> tableGlobalTableJoin = table.leftJoin(globalTable, keyMapper, joiner);
+        tableGlobalTableJoin.foreach(foreachAction);
+        runAndVerifyLeftJoin(inputTable);
+    }
+
+    @Test
+    public void shouldKTableGlobalKTableJoin() throws Exception {
+        final KTable<String, String> tableGlobalTableJoin = table.join(globalTable, keyMapper, joiner);
+        tableGlobalTableJoin.foreach(foreachAction);
+        runAndVerifyJoin(inputTable);
+    }
+
+    private void runAndVerifyJoin(final String topic) throws InterruptedException, ExecutionException {
```

Comment:
I've been looking a bit more and I'm still not certain what the correct thing to do in this situation is. We need to send an old value for the subtractors of the  aggregators/reducers, but the old value may not be the same value as previously seen. An example that doesn't work properly.
```
final GlobalKTable<String, String> g1 = builder.globalTable(Serdes.String(), Serdes.String(), "g1", "g1");
        final KTable<Integer, String> t1 = builder.table(Serdes.Integer(), Serdes.String(), "t1", "t1");
        t1.leftJoin(g1, (key, value) -> value, (v1, v2) -> v2)
           .groupBy((key, value) -> KeyValue.pair(value, value, Serdes.String(), Serdes.String()).count("count");
``` 
if i initialize g1 with:
(1, green)
(2, blue)
(3, yellow)
(4 red)
Then send to t1
(1, 1)
(2, 1)
(3, 1)
and flush state i get 
(green, 3)
all good so far.
However if i then send to g1
(1, orange)
and then to t1
(1, 4)
and flush state i get
(green, 3)
(orange, -1)
(red, 1)
which is obviously incorrect. The oldvalue of green never gets sent so the count for green doesn't reduce by 1, rather the oldvalue is orange, hence orange with a count of -1. HMMMM!